rsp_vision/save/save_data.py

```diff
@@ -0,0 +1,378 @@
+
+from rsp_vision.objects.enums import PhotonType
+from rsp_vision.objects.folder_naming_specs import FolderNamingSpecs
+from rsp_vision.objects.photon_data import PhotonData
+from rsp_vision.objects.SWC_Blueprint import (
+    SessionFolder,
+    SubjectFolder,
+    SWC_Blueprint_Spec,
+)
+
+
+def save_data(
+    swc_blueprint_spec: SWC_Blueprint_Spec,
+    folder_naming_specs: FolderNamingSpecs,
+    photon_data: PhotonData,
+    config_file: dict,
+) -> None:
+    """
+    This method saves the data in the appropriate folders, and updates the
+    analysis_log.csv file.
+
+    Parameters
+    ----------
+    swc_blueprint_spec : SWC_Blueprint_Spec
+        The object containing the information about the project
+        name and the path to the local folder where the data will be saved.
+    folder_naming_specs : FolderNamingSpecs
+        The object containing the information about the mouse line,
+        mouse id, folder name, and the path to the local folder where the
+        data will be saved.
+    photon_data : PhotonData
+        The object containing the data to be saved.
+    config_file : dict
+        The dictionary containing the information about the experimental
+        variables.
+    """
+    sub, ses, reanalysis, analysis_log = get_sub_and_ses(
+        folder_naming_specs, swc_blueprint_spec
+    )
+
+    subject_folder = SubjectFolder(
+        swc_blueprint_spec=swc_blueprint_spec,
+        folder_or_table=folder_naming_specs,
+        sub_num=sub,
+    )
+
+    session_folder = SessionFolder(
+        subject_folder=subject_folder,
+        folder_or_table=folder_naming_specs,
+        ses_num=ses,
+    )
+
+    Path(session_folder.ses_folder_path).mkdir(parents=True, exist_ok=True)
+
+    save_roi_info_and_fit_outputs(
+        photon_data=photon_data,
+        session_folder=session_folder,
+    )
+
+    save_signal_data_for_each_roi(
+        photon_data=photon_data,
+        session_folder=session_folder,
+    )
+
+    save_metadata_about_this_analysis(
+        folder_naming_specs=folder_naming_specs,
+        photon_data=photon_data,
+        session_folder=session_folder,
+        config_file=config_file,
+    )
+
+    save_info_in_main_log(
+        folder_naming_specs=folder_naming_specs,
+        photon_data=photon_data,
+        session_folder=session_folder,
+        analysis_log=analysis_log,
+        subject_folder=subject_folder,
+        reanalysis=reanalysis,
+        swc_blueprint_spec=swc_blueprint_spec,
+    )
+
+
+def get_sub_and_ses(
+    folder_naming_specs: FolderNamingSpecs,
+    swc_blueprint_spec: SWC_Blueprint_Spec,
+) -> tuple:
+    """
+    This method is charged with finding the subject and session numbers
+    for the current analysis. It also checks if the analysis is a reanalysis
+    of a previous one. If so, it returns the subject and session numbers
+    of the previous analysis.
+
+    There are 5 possible cases:
+    1. no data in the analysis.log file -> sub = 0, ses = 0
```

Comment:
This means the sub and ses are zero-indexed. In BIDS and SWC-Blueprint the convention for sub/ses IDs is to start with 1, not zero

Reply:
I see, thanks. But by keeping it 0-indexed I will match the 0-indexing structure of pandas dataframes (and python in general) 🤔
I'll think a little bit more about it

Reply:
I will keep 0-indexing for now